design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,129 @@
+the topics stored in Kafka as the source of truth. However, since there is no clear ownership, the defined table 
+could be mutated by using `INSERT` command to populate data into it. The insertion will pollute the original data source, 
+which was supposed to be owned by upstream such as a CDC pipeline that populates the topic.
+Secondly, when calling `CREATE TABLE`, the actual data is neither materialized nor ready for interactive queries to access. 
+To address these gaps, we would like to add:
+
+* Syntax to add read-only stream/table as `CREATE SOURCE STREAM/TABLE` to make the data source as read-only. Any 
+insertion command will be rejected, with an example error response as:
+
+```
+Error: insertion into source stream/table is not allowed
+```   
+
+* In addition, the `CREATE SOURCE TABLE` will become a persistent query instead of just a metadata operation like 
+`CREATE TABLE`, `CREATE STREAM`, or `CREATE SOURCE STREAM`. Source table will materialize the input topic data as 
+a RocksDB instance for interactive queries in the KsqlDB servers.
+
+## What is in scope
+* Add `CREATE SOURCE STREAM/TABLE` syntax to KSQL
+* `DESCRIBE ... EXTENDED` should show source streams and tables as read-only  
+* Add materialization of the `source table` to make it pull-query accessible
+* Make source stream/table immutable from KSQL by rejecting insertion requests
+ 
+
+## What is not in scope
+* We will not add role-based write permission for source stream/table in v1, even though it seems to be useful.
+
+## Design
+
+### Semantic Change
+We shall first add optional keyword `SOURCE` into the KSQL codebase, by adding the new syntax to `sqlBase.g4` for:
+* createStream
+* createTable
+
+with the sample syntax as:
+```
+CREATE (OR REPLACE)? (SOURCE)? TABLE (IF NOT EXISTS)?...
+```
+Then inside `AstBuilder#visitCreateStream` and `AstBuilder#visitCreateTable`, the built classes for stream/table 
+instances would be `CreateStream` and `CreateTable`, which are the only two subclasses of `CreateSource`. It makes sense 
+to add the `SOURCE` as a flag into `CreateSource`, and let `AstBuilder` pass the flag into the struct 
+for later stage reference of read-only.
+
+Additionally, when user calls `DESCRIBE ... EXTENDED`, the read-only attribute shall be displayed for stream/table. 
+
+### Make Source Stream/Table Read-only
+We plan to add restriction check into the `DistributingExecutor#throwIfInsertOnReadOnlyTopic` or create a similar function, 
+which could verify whether the given stream/table is read-only. If so, the insertion will be rejected.
+
+### Source Table Materialization
+In the call `EngineExecutor#execute`, we will check whether given plan has a `queryPlan` before deciding to proceed 
+creating a persistent query. To bypass this check, we need to provide a concrete query plan within the input plan. We would 
+inject a dummy query plan in `EngineExecutor#plan` phase, when the statement is a CreateTable statement with `SOURCE` tag.
```

Comment:
I don't think this is a "dummy" query plan - it's going to spin up a real query 😄 we should make sure to summarize the offline discussion about counting this toward the number of persistent queries

Reply:
> it's going to spin up a real query

I tend to disagree (at least from a terminology point of view). A "query" should have a `SELECT *` statements, but a `CT` and `CST` both don't have a SELECT but are DDL statements... Maybe we need to come up with a new term for this kind of KafkaStreams program that we deploy in the background.